# sfrugi_WG
 A study on copy number variations with intact open reading frames using individually assembled genomes

The scripts used in this study are grouped into the following directories, reflecting the original working environments:

Assembling:
 For the generation of individually assembled genomes.

RefSeq:
  Treatment of RefSeq data
  
GENE_ANNOTATIONS:
 Gene annotation for each genome assembly.

SNV:
 Identification of single nucleotide variations.

CNV_Gene:
 Identification of iORF-CNVs.

Ver7:
 Comparison of read depths between iORF-CNVs and other genes.

ABUNDANCE:
 Calculation of the abundance of iORF-CNVs and SNVs.

POSITION:
 Analysis of the genomic positions of iORF-CNV genes.

TRANSPOSITION:
 Testing retrotransposition as a potential origin of iORF-CNVs.

MALADAPTATION:
 Investigation of selective forces acting on iORF-CNV copies.

TRANSCRIPTION:
 Analysis of transcriptional mechanisms of iORF-CNVs.

SYNTHENY:
 Testing the conservation of synteny in potential iORF-CNVs.

TRANSPOSITION_short
  Testing if iORF-CNVs have higher proportion of intronless genes than non-iORF-CNVs for each length bin

GenomeSize
  Comparision of assembly sizes according to the sex

iORF_CNV_Coord
  listing iORF-CNVs
    
Quanti
  Quantification of iORF-CNVs with support for retrotransposition
  
    
drwxrwxr-x  4 kiwoong kiwoong  4096 déc.  16 10:59  TRANSPOSITION_short
drwxrwxr-x  6 kiwoong kiwoong  4096 déc.  16 16:33  
			

